### PR TITLE
feat: Add needed fields for imagecraft

### PR DESCRIPTION
**Pull request by upils**
- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox`?

-----

Implement the FO169 spec to adapt craft-archives to imagecraft.

I also documented the `priority` field to be closer to the documentation on snapcraft.io.

### CODE REVIEW

File: craft_archives/repo/package_repository.py [hunk 1]
```diff
@@ -277,6 +312,30 @@ def _not_mixing_suites_and_path(
...
 
+    @root_validator
+    def _not_mixing_suites_and_series_pocket(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        suites = values.get("suites")
+        series = values.get("series")
+        pocket = values.get("pocket")
+        url = values.get("url")
+        if suites and (series or pocket):
+            raise _create_validation_error(
+                url=url, message="specify either suites or series and pocket."
```
Comment (tigarmo) [inline craft_archives/repo/package_repository.py:323]: maybe we can make the message a bit clearer? Something like `suites cannot be used with series and pocket`

File: craft_archives/repo/apt_sources_manager.py
Comment (tigarmo) [inline craft_archives/repo/apt_sources_manager.py:1]: What about support for series and pocket when constructing the deb822 file? Is the idea to do that in a follow-up PR? I'd be fine with that

File: docs/reference/repo_properties.rst [hunk 2]
```diff
@@ -21,6 +21,31 @@ The following properties are supported for PPA-type repositories:
...
+     than 0
+   - Notes: string equivalencies are ``always``: 1000; ``prefer``: 990;
+     ``defer``: 100
+   - Examples:
+      - ``priority: always``
+      - ``priority: 1000``
+- key-id
+   - Type: string
+   - Description: 40 character GPG key identifier ("long-form thumbprint" or
+     "fingerprint"). The Craft Application will first look for the corresponding
```
Comment (tigarmo) [inline docs/reference/repo_properties.rst:38]: I think we need to drop this sentence referring to `<project>/snap/keys` from these docs - it only applies to snapcraft

File: docs/reference/repo_properties.rst [hunk 3]
```diff
@@ -21,6 +21,31 @@ The following properties are supported for PPA-type repositories:
...
+     key at: ``<project>/snap/keys/<key-id[-8:]>.asc``. If the key is not found,
+     the key will be fetched from the ``key-server``. To determine a key-id from
+     a given key file with gpg, type the following:
+     ``gpg --import-options show-only --import <file>``
+   - Format: alphanumeric, dash ``-`` , and underscores ``_`` permitted.
+   - Examples:
+       - ``key-id: 590CA3D8E4826565BE3200526A634116E00F4C82``
+
+         The Application will look for a corresponding key at
```
Comment (tigarmo) [inline docs/reference/repo_properties.rst:47]: here too

File: docs/reference/repo_properties.rst [hunk 4]
```diff
@@ -115,12 +157,21 @@ The following properties are supported for Deb-type repositories:
    - Examples:
        - ``suites: [xenial]``
        - ``suites: [xenial, xenial-updates]``
-- type
+- pocket
    - Type: enum[string]
-   - Description: Specifies type of package-repository
-   - Notes: Must be ``apt``
+   - Description: Pocket to get packages from, must one of: ``release``,
```
Comment (tigarmo) [inline docs/reference/repo_properties.rst:162]:
```suggestion
   - Description: Pocket to get packages from, must be one of: ``release``,
```

**Review by tigarmo — CHANGES_REQUESTED** [review]
thanks! This is very good, I added some mostly minor comments

File: craft_archives/repo/package_repository.py [hunk 5]
```diff
@@ -287,17 +346,18 @@ def _suites_without_backslash(cls, suite: str, values: Dict[str, Any]) -> str:
...
         url = values.get("url")
         if suites and not components:
             raise _create_validation_error(
                 url=url, message="components must be specified when using suites."
             )
-        if components and not suites:
+        if components and not (suites or pocket):
             raise _create_validation_error(
-                url=url, message="suites must be specified when using components."
+                url=url, message="suites or series and pocket must be specified when using components."
```
[the remark this answers is not in the document]
Comment (upils) [reply craft_archives/repo/package_repository.py:360]: You are right, this is confusing. I will fix that.

File: craft_archives/repo/apt_sources_manager.py
Comment (upils) [reply craft_archives/repo/apt_sources_manager.py:1]: Good catch, I forgot about it! Let me add it if this is not too big of a change and ask for another review.

Comment (upils) [reply craft_archives/repo/apt_sources_manager.py:1]: It is done. Let me know if this is what you had in mind. the `_get_suites()` function is generating a suites list tailored to what is expected in an official ubuntu image (so for example only `<series>-security` if the security pocket is given, etc.). Let me know if this is working for you. This is more opinionated than the implementation with the `suites` arg, where users can set whatever they want, even if it does make any sense "package-wise".

Comment (tigarmo) [reply craft_archives/repo/apt_sources_manager.py:1]: thanks! Do you happen to have a reference to the logic of which suites get used for which pockets?

**Comment by upils** [comment]
https://github.com/canonical/craft-archives/pull/120#discussion_r1662363606 This is the same logic we implemented in ubuntu-image (see https://github.com/canonical/ubuntu-image/blob/main/internal/imagedefinition/image_definition.go#L328). You can find documentation here https://canonical-ubuntu-packaging-guide.readthedocs-hosted.com/en/latest/explanation/archive/#pockets

File: craft_archives/repo/apt_sources_manager.py
Comment (upils) [reply craft_archives/repo/apt_sources_manager.py:1]: This is the same logic we implemented in ubuntu-image (see https://github.com/canonical/ubuntu-image/blob/main/internal/imagedefinition/image_definition.go#L328). You can find documentation here https://canonical-ubuntu-packaging-guide.readthedocs-hosted.com/en/latest/explanation/archive/#pockets

Comment (tigarmo) [reply craft_archives/repo/apt_sources_manager.py:1]: thanks. Do you think it's worth documenting this logic?

Comment (upils) [reply craft_archives/repo/apt_sources_manager.py:1]: I think we should avoid duplicating this documentation. But I can this link to the `pocket` field in the documentation to make it clear to users how this is supposed to behave.

Comment (tigarmo) [reply craft_archives/repo/apt_sources_manager.py:1]: Yes I'm fine with just adding a link

File: craft_archives/repo/apt_sources_manager.py [hunk 6]
```diff
@@ -26,6 +26,9 @@
 import distro
 
 from craft_archives import utils
+from craft_archives.repo.package_repository import (
+    PocketEnum,
+)
```
Comment (cmatsuoka) [inline craft_archives/repo/apt_sources_manager.py:31]: This could be a single line.

**Review by cmatsuoka — APPROVED** [review]

**Pull request merged**